File: backend/src/modules/transportation/costs/costs.service.ts
import { Injectable } from '@nestjs/common';
import * as ExcelJS from 'exceljs';
import { PrismaService } from '../../../prisma/prisma.service';
import { paginationHeader, success, successList } from '../../../common/response/base-response';

// Filters accepted by ReportCostOfLines (documented header filters, §6.11).
interface CostFilters {
  supplierId?: number;
  lineId?: number;
  serialBus?: string;
  dateFrom?: string;
  dateTo?: string;
}

export interface CostRow {
  LineName: string;
  Serial: string;
  NameOfRoute: string;
  SupplierName: string;
  LineCost: number;
  OneWay: boolean;
  CountOfround: number; // frozen odd-casing key
  DeductPerRound: number;
  DeductRoundNum: number;
  NetCost: number;
}

@Injectable()
export class CostsService {
  constructor(private readonly prisma: PrismaService) {}

  /** Build an inclusive [dateFrom, dateTo] day-range Prisma filter, or `undefined` (no bound). */
  private roundDateFilter(dateFrom?: string, dateTo?: string): Record<string, Date> | undefined {
    const bounds: Record<string, Date> = {};
    if (dateFrom) {
      const f = new Date(dateFrom);
      if (!isNaN(f.getTime())) bounds.gte = new Date(f.getFullYear(), f.getMonth(), f.getDate());
    }
    if (dateTo) {
      const t = new Date(dateTo);
      if (!isNaN(t.getTime())) {
        const lt = new Date(t.getFullYear(), t.getMonth(), t.getDate());
        lt.setDate(lt.getDate() + 1); // inclusive of the To calendar day
        bounds.lt = lt;
      }
    }
    return Object.keys(bounds).length > 0 ? bounds : undefined;
  }

  /**
   * Compute one TransportationReportCostData row for a route (§7.5 Cost report):
   *   two-way: rounds = checkIns + checkOuts ; NetCost = (LineCost/2)*rounds - deductions
   *   one-way: rounds = checkIns             ; NetCost =  LineCost   *rounds - deductions
   */
  private async costRowForRoute(
    r: { id: number; serial: string; nameOfRoute: string; lineCost: number; oneWay: boolean | null; line?: { lineName: string } | null; supplier?: { name: string } | null },
    dateCond: Record<string, Date> | undefined,
  ): Promise<CostRow> {
    const oneWay = r.oneWay ?? false;
    // Nullable round-date columns: with no date bound, still require the column present.
    const checkInCond = dateCond ?? { not: null };
    const checkOutCond = dateCond ?? { not: null };

    const [checkIns, checkOuts, deductions] = await Promise.all([
      this.prisma.transportationRoundForRoute.count({ where: { routeId: r.id, checkInDate: checkInCond } }),
      oneWay
        ? Promise.resolve(0)
        : this.prisma.transportationRoundForRoute.count({ where: { routeId: r.id, checkOutDate: checkOutCond } }),
      this.prisma.transportationVehicleRouteDeduction.findMany({
        where: { routeId: r.id, dateOfDeduction: dateCond },
        orderBy: { dateOfDeduction: 'desc' },
        select: { deductPerRound: true },
      }),
    ]);

    const countOfRound = oneWay ? checkIns : checkIns + checkOuts;
    const deductRoundNum = deductions.length;
    // Representative per-round deduction = the most recent row's value.
    const deductPerRound = deductRoundNum > 0 ? deductions[0].deductPerRound : 0;
    // Total actually deducted = sum of each round's deduction.
    const deductionsTotal = deductions.reduce((s, d) => s + d.deductPerRound, 0);
    const netCost = oneWay
      ? r.lineCost * countOfRound - deductionsTotal
      : (r.lineCost / 2) * countOfRound - deductionsTotal;

    return {
      LineName: r.line?.lineName ?? '',
      Serial: r.serial ?? '',
      NameOfRoute: r.nameOfRoute ?? '',
      SupplierName: r.supplier?.name ?? '',
      LineCost: r.lineCost,
      OneWay: oneWay,
      CountOfround: countOfRound,
      DeductPerRound: deductPerRound,
      DeductRoundNum: deductRoundNum,
      NetCost: netCost,
    };
  }

  /**
   * GET ReportCostOfLines — paginated TransportationReportCostData rows built
   * from routes (joined to line & supplier), with real round/deduction math (§7.5).
   */
  async reportCostOfLines(pageNo = 1, noOfItems = 20, filters: CostFilters = {}) {
    const where: Record<string, unknown> = { active: true };
    if (filters.lineId) where.transportationLineId = filters.lineId;
    if (filters.supplierId) where.supplierId = filters.supplierId;
    if (filters.serialBus) where.serial = filters.serialBus;

    const [total, rows] = await this.prisma.$transaction([
      this.prisma.transportationVehicleRoute.count({ where }),
      this.prisma.transportationVehicleRoute.findMany({
        where,
        include: { line: true, supplier: true },
        orderBy: { id: 'desc' },
        skip: (pageNo - 1) * noOfItems,
        take: noOfItems,
      }),
    ]);

    const dateCond = this.roundDateFilter(filters.dateFrom, filters.dateTo);
    const data = await Promise.all(rows.map((r) => this.costRowForRoute(r, dateCond)));

    return successList(data, paginationHeader(pageNo, noOfItems, total));
  }

  /** GET ReportCostOfLinesExcell — base64 (.xlsx) costs export (RTL, Arabic headers). */
  async reportCostOfLinesExcel(filters: CostFilters = {}) {
    const where: Record<string, unknown> = { active: true };
    if (filters.lineId) where.transportationLineId = filters.lineId;
    if (filters.supplierId) where.supplierId = filters.supplierId;
    if (filters.serialBus) where.serial = filters.serialBus;

    const rows = await this.prisma.transportationVehicleRoute.findMany({
      where,
      include: { line: true, supplier: true },
      orderBy: { id: 'desc' },
    });
    const dateCond = this.roundDateFilter(filters.dateFrom, filters.dateTo);
    const data = await Promise.all(rows.map((r) => this.costRowForRoute(r, dateCond)));

    const workbook = new ExcelJS.Workbook();
    const sheet = workbook.addWorksheet('تكلفة الخطوط');
    sheet.views = [{ rightToLeft: true }];
    sheet.columns = [
      { header: 'اسم الخط', key: 'line', width: 22 },
      { header: 'الرقم المسلسل', key: 'serial', width: 14 },
      { header: 'خط السير', key: 'route', width: 22 },
      { header: 'المورد', key: 'supplier', width: 22 },
      { header: 'تكلفة الخط', key: 'cost', width: 14 },
      { header: 'اتجاه واحد', key: 'oneWay', width: 12 },
      { header: 'عدد الجولات', key: 'rounds', width: 14 },
      { header: 'الخصم لكل جولة', key: 'deductPer', width: 16 },
      { header: 'عدد جولات الخصم', key: 'deductNum', width: 16 },
      { header: 'صافي التكلفة', key: 'net', width: 16 },
    ];
    sheet.getRow(1).font = { bold: true };
    for (const d of data) {
      sheet.addRow({
        line: d.LineName,
        serial: d.Serial,
        route: d.NameOfRoute,
        supplier: d.SupplierName,
        cost: d.LineCost,
        oneWay: d.OneWay ? 'نعم' : 'لا',
        rounds: d.CountOfround,
        deductPer: d.DeductPerRound,
        deductNum: d.DeductRoundNum,
        net: d.NetCost,
      });
    }

    const buffer = await workbook.xlsx.writeBuffer();
    return success(Buffer.from(buffer).toString('base64'));
  }
}

File: backend-dotnet/GarasTransport.Api/Modules/Transportation/Costs/CostsService.cs
using ClosedXML.Excel;
using GarasTransport.Api.Common;
using Microsoft.EntityFrameworkCore;

namespace GarasTransport.Api.Modules.Transportation.Costs;

// Filters accepted by ReportCostOfLines (documented header filters, §6.11).
public class CostFilters
{
    public int? SupplierId { get; set; }
    public int? LineId { get; set; }
    public string? SerialBus { get; set; }
    public string? DateFrom { get; set; }
    public string? DateTo { get; set; }
}

public class CostRow
{
    public string LineName { get; set; } = string.Empty;
    public string Serial { get; set; } = string.Empty;
    public string NameOfRoute { get; set; } = string.Empty;
    public string SupplierName { get; set; } = string.Empty;
    public double LineCost { get; set; }
    public bool OneWay { get; set; }
    public int CountOfround { get; set; } // frozen odd-casing key
    public double DeductPerRound { get; set; }
    public int DeductRoundNum { get; set; }
    public double NetCost { get; set; }
}

public class CostsService
{
    private readonly AppDbContext _db;

    public CostsService(AppDbContext db) => _db = db;

    /// <summary>Build an inclusive [dateFrom, dateTo] day-range filter, or (null, null) for no bound.</summary>
    private static (DateTime? Gte, DateTime? Lt) RoundDateFilter(string? dateFrom, string? dateTo)
    {
        DateTime? gte = null, lt = null;
        if (DateTime.TryParse(dateFrom, out var f)) gte = f.Date;
        if (DateTime.TryParse(dateTo, out var t)) lt = t.Date.AddDays(1); // inclusive of the To calendar day
        return (gte, lt);
    }

    /// <summary>
    /// Compute one TransportationReportCostData row for a route (§7.5 Cost report):
    ///   two-way: rounds = checkIns + checkOuts ; NetCost = (LineCost/2)*rounds - deductions
    ///   one-way: rounds = checkIns             ; NetCost =  LineCost   *rounds - deductions
    /// </summary>
    private async Task<CostRow> CostRowForRoute(TransportationVehicleRoute r, (DateTime? Gte, DateTime? Lt) dateCond)
    {
        var oneWay = r.OneWay ?? false;

        var checkInQuery = _db.TransportationRoundForRoutes.Where(x => x.RouteId == r.Id && x.CheckInDate != null);
        if (dateCond.Gte.HasValue) checkInQuery = checkInQuery.Where(x => x.CheckInDate >= dateCond.Gte.Value);
        if (dateCond.Lt.HasValue) checkInQuery = checkInQuery.Where(x => x.CheckInDate < dateCond.Lt.Value);
        var checkIns = await checkInQuery.CountAsync();

        var checkOuts = 0;
        if (!oneWay)
        {
            var checkOutQuery = _db.TransportationRoundForRoutes.Where(x => x.RouteId == r.Id && x.CheckOutDate != null);
            if (dateCond.Gte.HasValue) checkOutQuery = checkOutQuery.Where(x => x.CheckOutDate >= dateCond.Gte.Value);
            if (dateCond.Lt.HasValue) checkOutQuery = checkOutQuery.Where(x => x.CheckOutDate < dateCond.Lt.Value);
            checkOuts = await checkOutQuery.CountAsync();
        }

        var dedQuery = _db.TransportationVehicleRouteDeductions.Where(d => d.RouteId == r.Id);
        if (dateCond.Gte.HasValue) dedQuery = dedQuery.Where(d => d.DateOfDeduction >= dateCond.Gte.Value);
        if (dateCond.Lt.HasValue) dedQuery = dedQuery.Where(d => d.DateOfDeduction < dateCond.Lt.Value);
        var deductions = await dedQuery.OrderByDescending(d => d.DateOfDeduction).Select(d => d.DeductPerRound).ToListAsync();

        var countOfRound = oneWay ? checkIns : checkIns + checkOuts;
        var deductRoundNum = deductions.Count;
        var deductPerRound = deductRoundNum > 0 ? deductions[0] : 0; // most recent row's value
        var deductionsTotal = deductions.Sum();
        var netCost = oneWay
            ? r.LineCost * countOfRound - deductionsTotal
            : (r.LineCost / 2) * countOfRound - deductionsTotal;

        return new CostRow
        {
            LineName = r.Line?.LineName ?? string.Empty,
            Serial = r.Serial ?? string.Empty,
            NameOfRoute = r.NameOfRoute ?? string.Empty,
            SupplierName = r.Supplier?.Name ?? string.Empty,
            LineCost = r.LineCost,
            OneWay = oneWay,
            CountOfround = countOfRound,
            DeductPerRound = deductPerRound,
            DeductRoundNum = deductRoundNum,
            NetCost = netCost,
        };
    }

    private IQueryable<TransportationVehicleRoute> BuildQuery(CostFilters filters)
    {
        var q = _db.TransportationVehicleRoutes.Where(r => r.Active);
        if (filters.LineId.HasValue) q = q.Where(r => r.TransportationLineId == filters.LineId.Value);
        if (filters.SupplierId.HasValue) q = q.Where(r => r.SupplierId == filters.SupplierId.Value);
        if (!string.IsNullOrEmpty(filters.SerialBus)) q = q.Where(r => r.Serial == filters.SerialBus);
        return q;
    }

    /// <summary>GET ReportCostOfLines — paginated TransportationReportCostData rows.</summary>
    public async Task<object> ReportCostOfLines(int pageNo, int noOfItems, CostFilters filters)
    {
        var query = BuildQuery(filters);
        var total = await query.CountAsync();
        var rows = await query
            .Include(r => r.Line).Include(r => r.Supplier)
            .OrderByDescending(r => r.Id)
            .Skip((pageNo - 1) * noOfItems)
            .Take(noOfItems)
            .ToListAsync();

        var dateCond = RoundDateFilter(filters.DateFrom, filters.DateTo);
        var data = new List<CostRow>();
        foreach (var r in rows) data.Add(await CostRowForRoute(r, dateCond));

        return Envelope.SuccessList(data, Envelope.Pagination(pageNo, noOfItems, total));
    }

    /// <summary>GET ReportCostOfLinesExcell — base64 (.xlsx) costs export (RTL, Arabic headers).</summary>
    public async Task<object> ReportCostOfLinesExcel(CostFilters filters)
    {
        var rows = await BuildQuery(filters)
            .Include(r => r.Line).Include(r => r.Supplier)
            .OrderByDescending(r => r.Id)
            .ToListAsync();
        var dateCond = RoundDateFilter(filters.DateFrom, filters.DateTo);
        var data = new List<CostRow>();
        foreach (var r in rows) data.Add(await CostRowForRoute(r, dateCond));

        using var wb = new XLWorkbook();
        var sheet = wb.AddWorksheet("تكلفة الخطوط");
        sheet.RightToLeft = true;
        string[] headers = { "اسم الخط", "الرقم المسلسل", "خط السير", "المورد", "تكلفة الخط", "اتجاه واحد", "عدد الجولات", "الخصم لكل جولة", "عدد جولات الخصم", "صافي التكلفة" };
        for (var c = 0; c < headers.Length; c++) sheet.Cell(1, c + 1).Value = headers[c];
        sheet.Row(1).Style.Font.Bold = true;

        var row = 2;
        foreach (var d in data)
        {
            sheet.Cell(row, 1).Value = d.LineName;
            sheet.Cell(row, 2).Value = d.Serial;
            sheet.Cell(row, 3).Value = d.NameOfRoute;
            sheet.Cell(row, 4).Value = d.SupplierName;
            sheet.Cell(row, 5).Value = d.LineCost;
            sheet.Cell(row, 6).Value = d.OneWay ? "نعم" : "لا";
            sheet.Cell(row, 7).Value = d.CountOfround;
            sheet.Cell(row, 8).Value = d.DeductPerRound;
            sheet.Cell(row, 9).Value = d.DeductRoundNum;
            sheet.Cell(row, 10).Value = d.NetCost;
            row++;
        }

        return Envelope.Success(Fmt.WorkbookBase64(wb));
    }
}
